Signal/src/view controllers/AttachmentApprovalViewController.swift

```diff
@@ -0,0 +1,217 @@
+//
+
+import Foundation
+
+class AttachmentApprovalViewController: UIViewController {
+
+    let TAG = "[AttachmentApprovalViewController]"
+
+    // MARK: Properties
+
+    let attachment: SignalAttachment!
+
+    var successCompletion : (() -> Void)?
+
+    // MARK: Initializers
+
+    @available(*, deprecated, message:"use attachment: constructor instead.")
+    required init?(coder aDecoder: NSCoder) {
+        self.attachment = SignalAttachment.genericAttachment(withData: nil,
+                                                             dataUTI: kUTTypeContent as String)
+        super.init(coder: aDecoder)
+        assert(false)
+    }
+
+    required init(attachment: SignalAttachment!, successCompletion : @escaping () -> Void) {
+        assert(!attachment.hasError())
+        self.attachment = attachment
+        self.successCompletion = successCompletion
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    // MARK: View Lifecycle
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        view.backgroundColor = UIColor.black
+
+        self.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem:.done,
+            target:self,
+            action:#selector(donePressed))
+        self.navigationItem.title = NSLocalizedString("ATTACHMENT_APPROVAL_DIALOG_TITLE",
+                                                      comment: "Title for the 'attachment approval' dialog.")
+
+        createViews()
+    }
+
+    // MARK: - Create Views
+
+    private func createViews() {
+        let previewTopMargin = 30 as CGFloat
+        let previewHMargin = 20 as CGFloat
+
+        let attachmentPreviewView = UIView()
+        self.view.addSubview(attachmentPreviewView)
+        attachmentPreviewView.autoPinWidthToSuperview(withMargin:previewHMargin)
+        attachmentPreviewView.autoPin(toTopLayoutGuideOf: self, withInset:previewTopMargin)
+
+        createButtonRow(attachmentPreviewView:attachmentPreviewView)
+
+        if attachment.isImage() {
+            createImagePreview(attachmentPreviewView:attachmentPreviewView)
+        } else {
+            createGenericPreview(attachmentPreviewView:attachmentPreviewView)
+        }
+    }
+
+    private func createImagePreview(attachmentPreviewView: UIView) {
+        var image = attachment.image
+        if image == nil {
+            image = UIImage(data:attachment.data)
+        }
+        if image != nil {
+            let imageView = UIImageView(image:image)
+            imageView.layer.minificationFilter = kCAFilterTrilinear;
+            imageView.layer.magnificationFilter = kCAFilterTrilinear;
+            imageView.contentMode = .scaleAspectFit
+            attachmentPreviewView.addSubview(imageView)
+            imageView.autoPinWidthToSuperview()
+            imageView.autoPinHeightToSuperview()
+        } else {
+            createGenericPreview(attachmentPreviewView:attachmentPreviewView)
```

Comment:
If we can't load the image, show the "generic" preview.  This should never happen (thanks to the validation logic in `SignalAttachment`) but we want to fail gracefully if it does.